src/Payments.sol

```diff
@@ -392,49 +534,192 @@ contract Payments is
+        
+        return remainingSettlementEpochs;
+    }
+
+    function validateRateChangeRequirements(
+        Rail storage rail,
+        Account storage payer,
+        uint256 oldRate,
+        uint256 newRate,
+        uint256 lockupSettledUpto
+    ) internal view {
+        // For terminated rails, rate can only be decreased
+        require(
+            rail.terminationEpoch == 0 || newRate <= oldRate,
+            "failed to modify rail: cannot increase rate on terminated rail"
+        );
+
+        // Case 1: Lockup fully settled -> all good
+        if (lockupSettledUpto == block.number) {
+            return;
+        }
+        
+        // Case 2.A: Lockup not fully settled -> check if rail is in debt
+        if (block.number > payer.lockupLastSettledAt + rail.lockupPeriod) {
+            require(
+                newRate == oldRate,
+                "rail is in-debt; cannot change rate"
+            );
+            return;
+        }
+        
+        // Case 2.B: Lockup not fully settled -> check if rate is being increased
+        require(
+            newRate <= oldRate,
+            "account lockup not fully settled; cannot increase rate"
+        );
+    }
+
+    function validateAndUpdateLockupAllowance(
+        OperatorApproval storage approval,
+        uint256 oldLockup,
+        uint256 newLockup
+    ) internal {
+        // Check if we're increasing or decreasing lockup
+        if (newLockup >= oldLockup) {
+            uint256 lockupIncrease = newLockup - oldLockup;
+
+            // Only check allowance if we're increasing lockup
+            require(
+                approval.lockupUsage + lockupIncrease <= approval.lockupAllowance,
+                "exceeds operator lockup allowance"
+            );
+
+            // Update usage
+            approval.lockupUsage += lockupIncrease;
+        } else if (newLockup < oldLockup) {
+            // this should never underflow as lockupUsage should always be greater than the `oldLockup` 
+            uint256 lockupDecrease = oldLockup - newLockup;
+            approval.lockupUsage -= lockupDecrease;
+        }
+    }
 
     function validateAndModifyRateChangeApproval(
         Rail storage rail,
         OperatorApproval storage approval,
         uint256 oldRate,
         uint256 newRate,
-        uint256 oneTimePayment
+        uint256 oneTimePayment,
+        uint256 effectiveLockupPeriod
     ) internal {
         // Ensure the one-time payment does not exceed the available fixed lockup on the rail.
         require(
             oneTimePayment <= rail.lockupFixed,
             "one-time payment exceeds rail fixed lockup"
         );
 
-        // Calculate the original total lockup amount:
         uint256 oldTotalLockup = (oldRate * rail.lockupPeriod) +
             rail.lockupFixed;
-        uint256 newTotalLockup = (newRate * rail.lockupPeriod) +
+        uint256 newTotalLockup = (newRate * effectiveLockupPeriod) +
             rail.lockupFixed;
```

Comment:
@Stebalien I am not sure how to calculate the old lockup here

Reply:
The calculation should be the same as the new `effectiveLockupPeriod`, just with the old period in case the period changed. That is, it's the period minus the time between the termination epoch and the current epoch (if any).

That assumes lockup was fully up-to-date at termination but... I think we should stick with that for now.

Reply:
We do now ensure that the account lockup is fully settled during rail termination. It's a reasonable ask from the user and makes a whole host of things easy to reason about.